<launch>
     <!-- saving trajectory path and timing information -->
        <arg name="dosave"      default="true" />
        <arg name="dotime"      default="true" />
        <arg name="path_rect_est"    default="/tmp/traj_rect_estimate.txt" />
        <arg name="path_est"    default="/tmp/traj_estimate.txt" />
        <arg name="path_time"   default="/tmp/traj_timing.txt" />

     <node type="rviz" name="rviz" pkg="rviz" args="-d $(find ov_msckf)/launch/display_4.rviz" />
    <!-- MASTER NODE! -->
    <node name="run_serial_msckf" pkg="ov_msckf" type="run_serial_msckf" output="screen" clear_params="true" required="true">
<!--        <node name="run_serial_msckf" pkg="ov_msckf" type="run_serial_msckf" output="screen" clear_params="true" required="true" launch-prefix="gdb -ex run &#45;&#45;args">-->

        <!-- bag topics -->
        <param name="topic_imu"      type="string" value="/imu" />
        <param name="topic_camera0"  type="string" value="/img0" />
        <param name="topic_camera1"  type="string" value="/img1" />

        <!-- bag parameters -->

        <!-- <param name="path_bag"    type="string" value="/home/zzq/DataSet/euroc_mav/V1_01_easy.bag" /> -->
        <!-- <param name="path_bag"    type="string" value="/home/zzq/DataSet/euroc_mav/V1_02_medium.bag" /> -->
         <!-- <param name="path_bag"    type="string" value="/home/zzq/DataSet/euroc_mav/V1_03_difficult.bag" /> -->
         <!-- <param name="path_bag"    type="string" value="/home/zzq/DataSet/euroc_mav/V2_01_easy.bag" /> -->
         <!-- <param name="path_bag"    type="string" value="/home/zzq/DataSet/332.97713134460906uroc_mav/MH_01_easy.bag" /> -->
        <!-- <param name="path_bag"    type="string" value="/home/zzq/DataSet/euroc_mav/MH_02_easy.bag" /> -->
        <!-- <param name="path_bag"    type="string" value="/home/zzq/DataSet/euroc_mav/MH_03_medium.bag" /> -->
        <!-- <param name="path_bag"    type="string" value="/home/zzq/DataSet/euroc_mav/MH_04_difficult.bag" /> -->
        <param name="path_bag" type="string" value="/media/zzq/BackupDrive/DataSet/4seasons/office_loop/office_loop_2/office_loop_2.bag" />
        


       <!-- <param name="path_gt"     type="string" value="$(find ov_data)/euroc_mav/MH_02_easy.txt" /> -->
       <!-- <param name="path_gt"     type="string" value="$(find ov_data)/euroc_mav/MH_03_medium.txt" /> -->
       <!-- <param name="path_gt"     type="string" value="$(find ov_data)/euroc_mav/MH_04_difficult.txt" /> -->
        <param name="path_gt"         type="string" value="$(find ov_data)/4seasons/office_loop_2.txt" />
<!--        <param name="path_bag"    type="string" value="/home/patrick/datasets/eth/V2_03_difficult.bag" />-->
<!--        <param name="path_gt"     type="string" value="$(find ov_data)/euroc_mav/V2_03_difficult.csv" />-->
        <param name="bag_start"   type="double" value="10.0" />   <!-- v1-2: 0, mh1: 40, mh2: 35, mh3-5: 20 -->
        <param name="bag_durr"    type="int"    value="-1" /> 


         <!-- match map parameters-->
       <param name="use_prior_map" type="bool" value="false" />
        <param name="map_save_path" type="string" value="/home/zzq/Code/map_open_vins_v2_ros/src/open_vins/matches/" />
        <!-- <param name="pose_graph_filename" type="string" value="0827loc_timeq_timer_2dq_2dr_3dr.txt" /> -->
        <!-- <param name="pose_graph_filename" type="string" value="0828loc_timeq_timer_2dq_2dr_3dr.txt" /> -->
        <!-- <param name="pose_graph_filename" type="string" value="0129loc_timeq_timer_2dq_2dr_3dr.txt" /> -->
        <param name="pose_graph_filename" type="string" value="MH_05_loc_timeq_timer_2dq_2dr_3dr.txt" />
        <!-- <param name="path_gt"     type="string" value="/home/zzq/Code/map_open_vins_ros/src/open_vins/matches/gt_imu_20170828.txt" /> -->
        <param name="keyframe_pose_filename" type="string" value="MH_01_easy_cam.txt" />
        <param name="points_filename" type="string" value="/tmp/match_points.txt" />
        <param name="save_match_points" type="bool" value="true" />
         <param name="save_match_points" type="bool" value="true" />
        <param name="match_num_thred" type="int" value="20" />
        <param name="pixel_error" type="double" value="3.0" />
        <param name="transform_cov" type="double" value="20.0" />
        <param name="keyframe_cov_pos" type="double" value="0.11"/>
        <param name="keyframe_cov_ori" type="double" value="0.11"/>

        <param name="iter_num" type="int" value="30"/>
        <param name="opt_thred" type="double" value="1"/>

        <param name="kf_match_interval" type="double" value="0" />
        <param name="use_robust_match" type="bool" value="true" />


        <!-- world/filter parameters -->
        <param name="use_fej"                type="bool"   value="true" />
        <param name="use_imuavg"             type="bool"   value="true" />
        <param name="use_rk4int"             type="bool"   value="true" />
        <param name="use_stereo"             type="bool"   value="true" />
        <param name="calib_cam_extrinsics"   type="bool"   value="true" />
        <param name="calib_cam_intrinsics"   type="bool"   value="true" />
        <param name="calib_cam_timeoffset"   type="bool"   value="true" />
        <param name="calib_camimu_dt"        type="double" value="0.0" />
        <param name="max_clones"             type="int"    value="11" />
        <param name="max_slam"               type="int"    value="75" />
        <param name="max_slam_in_update"     type="int"    value="25" /> <!-- 25 seems to work well -->
        <param name="max_msckf_in_update"    type="int"    value="999" />
        <param name="max_cameras"            type="int"    value="2" />
        <param name="dt_slam_delay"          type="double" value="5" />
        <param name="init_window_time"       type="double" value="0.75"  />
        <param name="init_imu_thresh"        type="double" value="0.2" />
        <rosparam param="gravity">[0.0,0.0,9.81]</rosparam>
        <param name="feat_rep_msckf"         type="string" value="GLOBAL_3D" />
        <param name="feat_rep_slam"          type="string" value="ANCHORED_FULL_INVERSE_DEPTH" />
        <param name="feat_rep_aruco"         type="string" value="ANCHORED_FULL_INVERSE_DEPTH" />

        <!-- zero velocity update parameters -->
        <param name="try_zupt"               type="bool"   value="false" />
        <param name="zupt_chi2_multipler"    type="int"    value="2" />
        <param name="zupt_max_velocity"      type="double" value="0.3" />
        <param name="zupt_noise_multiplier"  type="double" value="50" />

        <!-- timing statistics recording -->
        <param name="record_timing_information"   type="bool"   value="false" />
        <param name="record_timing_filepath"      type="string" value="/tmp/timing_stereo.txt" />

        <!-- tracker/extractor properties -->
        <param name="use_klt"            type="bool"   value="true" />
        <param name="num_pts"            type="int"    value="300" />
        <param name="fast_threshold"     type="int"    value="15" />
        <param name="grid_x"             type="int"    value="5" />
        <param name="grid_y"             type="int"    value="3" />
        <param name="min_px_dist"        type="int"    value="10" />
        <param name="knn_ratio"          type="double" value="0.70" />
        <param name="downsample_cameras" type="bool"   value="false" />

        <!-- aruco tag/mapping properties -->
        <param name="use_aruco"        type="bool"   value="false" />
        <param name="num_aruco"        type="int"    value="1024" />
        <param name="downsize_aruco"   type="bool"   value="true" />

        <!-- sensor noise values / update -->
        <param name="up_msckf_sigma_px"            type="double"   value="1" />
        <param name="up_msckf_chi2_multipler"      type="double"   value="1" />
        <param name="up_slam_sigma_px"             type="double"   value="1" />
        <param name="up_slam_chi2_multipler"       type="double"   value="1" />
        <param name="up_aruco_sigma_px"            type="double"   value="1" />
        <param name="up_aruco_chi2_multipler"      type="double"   value="1" />
        <param name="gyroscope_noise_density"      type="double"   value="1.6968e-04" />
        <param name="gyroscope_random_walk"        type="double"   value="1.9393e-05" />
        <param name="accelerometer_noise_density"  type="double"   value="2.0000e-3" />
        <param name="accelerometer_random_walk"    type="double"   value="3.0000e-3" />
<!--         <param name="gyroscope_noise_density"      type="double"   value="0.005" /> -->
<!--                 <param name="gyroscope_random_walk"        type="double"   value="0.001" /> -->
<!--                 <param name="accelerometer_noise_density"  type="double"   value="0.05" /> -->
<!--                 <param name="accelerometer_random_walk"    type="double"   value="0.01" /> -->


        <!-- camera intrinsics -->
         <param name="gyroscope_noise_density"      type="double"   value="0.000187 " />
        <param name="gyroscope_random_walk"        type="double"   value="2.66e-05" />
        <param name="accelerometer_noise_density"  type="double"   value="0.00186" />
        <param name="accelerometer_random_walk"    type="double"   value="0.000433" />

        <!-- camera intrinsics -->
        <rosparam param="cam0_wh">[800, 400]</rosparam>   <!-- rosparam以yaml来读数据，通常用于矩阵的读取-->
        <rosparam param="cam1_wh">[800, 400]</rosparam>
        <param name="cam0_is_fisheye" type="bool" value="false" />
        <param name="cam1_is_fisheye" type="bool" value="false" />
        <rosparam param="cam0_k">[527.9990706330082, 527.963495807245, 399.18451401412665, 172.8193108347693]</rosparam>
        <rosparam param="cam0_d">[-0.03559759964255725, -0.005093721310999416, 0.019716282737702494, -0.01583280039499382]</rosparam>
        <rosparam param="cam1_k">[529.2496538273606, 529.4013679656194, 412.4733148308946, 172.1405434152354]</rosparam>
        <rosparam param="cam1_d">[-0.03696481607810142, 0.0102400942942071, -0.01632902510720569, 0.009597717025035516]</rosparam>

        <!-- camera extrinsics -->
        <rosparam param="T_C0toI">
            [
             -0.999885, -0.00689016,   0.0134931, 0.176201,
             -0.013523,  0.00430464,   -0.999899, -0.0557441,
              0.00683139,   -0.999967, -0.00439732, 0.00223549,
            0.0, 0.0, 0.0, 1.0
            ]
        </rosparam>
        <rosparam param="T_C1toI">
            [
            -0.999995,  0.00317347, 0.000428064, -0.124312,
            -0.00041104,  0.00536298,   -0.999986, -0.0566834,
            -0.00317572,   -0.999981, -0.00536165, 0.00164873,
            0.0, 0.0, 0.0, 1.0
            ]
        </rosparam>

    </node>

  <!-- record the trajectory if enabled -->
  <group if="$(arg dosave)">
          <node name="recorder_estimate" pkg="ov_eval" type="pose_to_file" output="screen" required="true">
           <param name="topic"      type="str" value="/ov_msckf/poseimu" />
            <param name="topic_type" type="str" value="PoseWithCovarianceStamped" />
            <param name="output"     type="str" value="$(arg path_est)" />
            <param name="topic_rect"      type="str" value="/ov_msckf/poseimu_rect" />
            <param name="topic_type_rect" type="str" value="PoseWithCovarianceStamped" />
            <param name="output_rect"     type="str" value="$(arg path_rect_est)" />
          </node>
      </group>

</launch>